### ISSUE TITLE
Add Hijri & Hebrew Calendar support to datepicker

### ISSUE THREAD

**Issue by icfantv** [ng-bootstrap/ng-bootstrap#656]
The datepicker needs to have Hijri Calendar support.  We should not support the other three Arabic calendars.
- The Hijri Calendar has days, weeks, months, and years.
- The new year comes 11 days earlier every year compared to the Gregorian Calendar so there's a shift over time.
- There are the same number of months as the Gregorian Calendar
- The Higri Calendar reads RTL with the Arabic equivalent of Saturday being the first day of the week.
- We need abbreviated days of the week.
- We need abbreviated months of the year.

**Comment by icfantv** [comment]
@AhmedMustafa, I know this was a request of yours for UIBS and we recognize the importance of getting it into the ng2 port so we would love your feedback and assistance here.  Can you please review the issue bullets and add, via comments, anything that would help correct/clarify?  Thank you.  E.g.,

Is it more proper to utilize Arabic abbreviations of days of the week and month names or use abbreviations of transliterated names?

**Comment by wesleycho** [comment]
Is there a reason we don't want to support the other calendars? Is there something special about them?

**Comment by icfantv** [comment]
@wesleycho the way our i18n (Localization) department explained it to me was that the bulk of the Arab world watches one channel (station?) called MBC and this is the calendar format that they (the TV station) use.  If anyone else chimes in with a _legitimate_ reason to be supporting the other calendars, it may be worth revisiting.  But for now, I say just stick with the Hijri calendar.

**Comment by AhmedMustafa** [comment]
Hijri is a general term used to describe the Islamic lunar-based calendar. There is no specific Islamic calendar named Hijri (all are Hijri calendar).
However, there are many types of Hijri/Islamic calendars used in the Islamic countries ([Calendar Types](http://cldr.unicode.org/development/development-process/design-proposals/islamic-calendar-types) ). All have the following characteristics:
- All have days, weeks, months, and years.
- All have a fixed 12 months per year
- Each month has 29/30 days (no more, no less)
- Islamic year is about 354/355 days
- The first day of the week differs from country to another. It is a country-related not calendar-related property.

**Islamic(arithmetic), Islamic-civil, Islamic-tabular & Islamic-umalqura** all are different types of Hijri calendar. Many of them are implemented in a plenty of technologies such as Java, Dojo, ibm-js, jQuery, Globalize,..etc. Each calendar is using a different type of calculations (Astronomical, Table-Based,...). 
My team (Arabic Competence & Globalization Center @ IBM Egypt) has made many contributions to those technologies & we are welling to contribute here as well.

Here is some examples of our previous work in different libraries.
- [umalquraCalendar.js in ibm-js/ecma402](https://github.com/ibm-js/ecma402/blob/master/calendars/umalquraCalendar.js)
- [islamicCalendar.js in ibm-js/ecma402](https://github.com/ibm-js/ecma402/blob/master/calendars/islamicCalendar.js)
- [civilCalendar.js in ibm-js/ecma402](https://github.com/ibm-js/ecma402/blob/master/calendars/civilTabularCalendar.js)
- [islamicCalendar.js in Dojo](https://github.com/dojo/dojox/blob/master/date/islamic/Date.js)
- [umalquraCalendar.js in Dojo](https://github.com/dojo/dojox/blob/master/date/umalqura/Date.js)

if you have a design that separate the UI from the core calculations modules, it will be easy to plug any calendar implementation (Hijri, Hebrew, Buddhist, Persian,...etc) seamlessly.


> Is it more proper to utilize Arabic abbreviations of days of the week and month names or use abbreviations of transliterated names?

I think it is better to rely on the CLDR data for the Arabic/English,... abbreviations.
[cldr-cal-islamic-full/main/ar/ca-islamic.json] (https://github.com/unicode-cldr/cldr-cal-islamic-full/blob/master/main/ar/ca-islamic.json)
[cldr-cal-islamic-full/main/en/ca-islamic.json] (https://github.com/unicode-cldr/cldr-cal-islamic-full/blob/master/main/en/ca-islamic.json)

**Comment by icfantv** [comment]
@AhmedMustafa this is good stuff.  thank you!

From what I can see, it appears to be abstracted/extensible, but I don't know if the base `NgbCalendar` class has everything needed to wire in a new calendar.  You can take a look [at this](https://github.com/ng-bootstrap/ng-bootstrap/blob/master/src/datepicker/ngb-calendar.ts) to see what @maxokorokov did for our first calendar commit (which was this past week).

I'm fine using that JSON file as the "source," but it would be ideal if we could reference it at build time and pull in the proper values.  Do you happen to know if there's an NPM package that has this data?  I found [this cldr package](https://www.npmjs.com/package/cldr) but will defer to your expertise here as to which NPM package we could/should use.

Monday is a holiday in the US so let's try and touch base with @pkozlowski-opensource and @maxokorokov and see about how best to proceed.  In the interim, if you can look at the link i posted above and see if/how the design might need to be changed to accommodate other calendars (specifically Islamic), please do so.  Thanks again.

**Comment by AhmedMustafa** [comment]
It seems good. Great job @maxokorokov.
However, in order to accommodate the Islamic/Hebrew calendars, I think we should extend the NgbDate class to override setMonth & setDay methods. The new NgbDateIslamic/NgbDateHebrew classes should have methods to convert to/from Gregorian date plus some other helper methods. Thus, I think fromJSDate, toJSDate should be part of the NgbDate.

It is very important to have fromJSDate, toJSDate inside NgbDateIslamic/NgbDateHebrew classes as we expect that all the dates provided to the calendar to be JS dates (Gregorian). And it is also expected that the the final value submitted to the backend to be Gregorian date as well.

> Do you happen to know if there's an NPM package that has this data?

The only package I know is [cldrjs ](https://www.npmjs.com/package/cldrjs) used by [Globalize ](https://github.com/jquery/globalize/).
It might be not so bad if we start with our own localiztion module that provides months and days abbreviations in Arabic, Hebrew and English for the Islamic and Hebrew calendar.

Is there any expected release date for this feature?

**Comment by AhmedMustafa** [comment]
I can help with the implementation of the IslamicDate/IslamicCalendar modules if it is possible.

**Comment by AhmedMustafa** [comment]
@tomerm FYI

**Comment by tomerm** [comment]
@AhmedMustafa thanks for sharing. We can certainly join the party and provide implementation of Hebrew calendar along the same lines as you pointed out above. 
@ashensis FYI.

**Comment by pkozlowski-opensource** [comment]
@AhmedMustafa @tomerm awesome guys, please do contribute!  IMO the most important part right now is that you start playing with a potential impl to see if the API we've got in place and all the flex points are enough to implement other calendars. 

As a reminder, here is the main abstractions we've got in place right now: 
* `NgbCalendar`: https://github.com/ng-bootstrap/ng-bootstrap/blob/a729ba981f72c38287da0033dd5344e7b2425c98/src/datepicker/ngb-calendar.ts#L14-L26
* `NgbDatepickerI18n` (names, i18n): https://github.com/ng-bootstrap/ng-bootstrap/blob/a729ba981f72c38287da0033dd5344e7b2425c98/src/datepicker/datepicker-i18n.ts#L11-L23

if any flex points are to be added it would be great to know it sooner than later :-)

**Comment by maxokorokov** [comment]
@AhmedMustafa thanks. However current implementation is a bit different from what you were saying. It's not the `NgbDate` that should be extended, but the `NgbCalendar`. 

Now the `NgbDate` is just an internal data model to hold year, month and day and not use JS date internally. JS Date is only used internally by the default `NgbCalendarGregorian` to do the calculations.

So it should be more like new `NgbCalendarIslamic` and `NgbCalendatHebrew` that use `NgbDate`.

**Comment by maxokorokov** [comment]
Hey, @AhmedMustafa, apologies for the delay. I started to look at the pull request and calendar integration again. It looks terrific, thanks a lot for spending time and understanding how the current implementation works!

I played with the demo and calendar implementation and almost just worked for me!
I have a couple of questions:

1. The opened pull request is missing i18n part, so for now the months are displayed a `0=Jan, 1=Feb`, etc. I guess should be `0=Muharram, 1=Safar`, etc for english. It will be very easy to add (and can always be added later separately), so the question is: I can't find on the internet if there are common abberviations for month names, like `January → Jan`, `Muharram → ?`, etc.

2. I guess the week always starts with `Sunday` and it doesn't make sense to allow changing that?

3. What is the common weekday numbering, like `Sun=0, Mon=1`?

4. Is it common to start month numbering with `0-11`, not `1-12`? Asking because we've just changed it recently for the default Gregorian one.

Will go through the pull request once more a little later and leave comments there.

**Comment by icfantv** [comment]
@maxokorokov i think the abbreviations you want [are here](https://github.com/unicode-cldr/cldr-cal-islamic-full/blob/master/main/ar/ca-islamic.json).  That's the CLDR that was mentioned above.  We should, at some point, investigate being able to pull in the i18n stuff from reliable third parties so we don't have to manage and store it ourselves.  That said however, it's not like the abbreviations change...

**Comment by icfantv** [comment]
I suppose the other decision is whether or not to use the transliterated spelling or the actual spelling in the proper language.  I'm checking with our i18n dept. to see what is more common.

**Comment by maxokorokov** [comment]
@icfantv thanks for the links, that's what I was looking for

Personally I would expect to have transliterated values if the site language is english. Mac OS does that for instance, if you change just the calendar, but not the language

**Comment by icfantv** [comment]
@maxokorokov I spoke with our i18n department and there are no non-transliterated abbreviations, at least in Arabic, for month and weekday names.  They told me that to abbreviate it would be very weird and not normal.  The abbreviated, transliterated values (at least in Arabic) are are able to be discerned, but I got the sense that they, too, are not "normal."  I'm not keen on having a configuration option for as to whether or not the week headings (and month name) should use the transliterated values, but this may be necessary as we can't really guess what the end-user is going to want.  @pkozlowski-opensource do you have any thoughts here?

@AhmedMustafa can you weigh in here, please?  What would be the ideal configuration from an Arabic perspective?  Thanks.

**Comment by AhmedMustafa** [comment]
@maxokorokov 

> The opened pull request is missing i18n part.

I managed to handle the i18n part later. I just wanted to know your feedback on the approach I followed before going further with the implementation. 

> I can't find on the internet if there are common abberviations for month names, like January → Jan, Muharram → ?, etc.

In my opinion, the [unicode-cldr](https://github.com/unicode-cldr) is the most reliable source to use.

> I guess the week always starts with Sunday and it doesn't make sense to allow changing that?

The first day of the week differs from country to another. However, I think it is not critical for the end-user to have the calendar start on Saturday or Sunday. So, I think it is safe to skip that feature until we get a real business case.

>  - What is the common weekday numbering, like Sun=0, Mon=1

> -  Is it common to start month numbering with 0-11, not 1-12? 

From end-user point of view, it is more common & less confusing to start numbering from 1.

**Comment by AhmedMustafa** [comment]
@icfantv 

> can you weigh in here, please? What would be the ideal configuration from an Arabic perspective?

That is also differs from country to another & again the most reliable source to follow is the [unicode-cldr](https://github.com/unicode-cldr).
For example, the [Lebanese ](https://github.com/unicode-cldr/cldr-dates-full/blob/master/main/ar-LB/ca-gregorian.json)users expect to see different names/abbreviations for months other than what the [Egyptian ](https://github.com/unicode-cldr/cldr-dates-full/blob/master/main/ar-EG/ca-gregorian.json)users expect to see.
However, the common names/abbreviations used in the Arabic countries in both Arabic and English languages, for both Islamic and Gregorian calendars, is as follows:
[Gregorian-En](https://github.com/unicode-cldr/cldr-dates-full/blob/master/main/en/ca-gregorian.json)
[Gregorian-Ar](https://github.com/unicode-cldr/cldr-dates-full/blob/master/main/ar/ca-gregorian.json)
[Islamic-En](https://github.com/unicode-cldr/cldr-cal-islamic-full/blob/master/main/en/ca-islamic-civil.json)
[Islamic-Ar](https://github.com/unicode-cldr/cldr-cal-islamic-full/blob/master/main/ar/ca-islamic-civil.json)

**Comment by icfantv** [comment]
@AhmedMustafa sorry for the confusion, my question wasn't what the applicable Gregorian names were, but rather what the appropriate solution was for the Arabic names.  Our i18n department indicated that Arabic users are not used to seeing abbreviations in Arabic, but it _might_ work for transliterated Arabic - though it would be weird.

My question was centered around whether or not it made sense to make this configurable for the end user.  The other thing I've noticed is that many Arabic calendar widgets show _both_ the Arabic and Gregorian names and numbering systems (with one being "primary").

Was that more clear?  Thanks.

**Comment by AhmedMustafa** [comment]
I became more confused :) , So it will be great if you provide some examples so that I may be sure what you are exactly mean.

> what the appropriate solution was for the Arabic names. 

What do mean by `Arabic names `? days or months? 

> Our i18n department indicated that Arabic users are not used to seeing abbreviations in Arabic, but it might work for transliterated Arabic 

What do you mean by `abbreviations in Arabic` &  `transliterated Arabic`?  please provide some examples.

>  many Arabic calendar widgets show both the Arabic and Gregorian names 

What do you mean by `Arabic calendar`? Do you mean `Islamic calendar` or `Gregorian calendar translated to Arabic`? 
What do you mean by `Arabic and Gregorian names`? please provide some examples.

**Comment by icfantv** [comment]
@AhmedMustafa hah.  sorry.  ok.  here we go:

When I say Gregorian names I mean both: Monday (Mon) and January (Jan) and when I say Arabic names, I also mean both: محرم, Muharram (Muh) and الأحد, Al ‘ahad (not sure if/what the abbreviation would be, if one exists for the days of the week in transliterated form).  Was that more clear?

As far as the combined calendar, this is what I mean (the left side):
![arabic-hijri](https://cloud.githubusercontent.com/assets/370054/19421466/d4e5081a-93b6-11e6-98e7-af5db01cc67d.jpg)

The examples I've seen show both situations where the Arabic is "primary," meaning the larger text and the Gregorian as the smaller and vice versa.  I'm not saying we need to support this (or that we should), rather, just something else to think about.

Please let me know if that was better.

**Comment by AhmedMustafa** [comment]
> Al ‘ahad (not sure if/what the abbreviation would be, if one exists for the days of the week in transliterated form).

 I've never seen it before and I doubt it might be used anywhere in the Arabic countries.

This a list of the months/days names/abbreviations that is commonly used in the Arabic countries.
![gre_cal](https://cloud.githubusercontent.com/assets/7556286/19429542/812498e0-9450-11e6-8f8b-8bcfd77c5ee2.png)

![isl_cal](https://cloud.githubusercontent.com/assets/7556286/19429556/8e09a762-9450-11e6-95ce-1484b73fe3c1.png)

![weekday](https://cloud.githubusercontent.com/assets/7556286/19429563/962ffc20-9450-11e6-9d38-72e31eb18375.png)

From the tables above we can conclude the following:
 - In general, the abbreviations (months/days) are rarely used in any official documents, books or newspapers,..etc. While it is common to be used in smartphones, computers, watches...etc. 
 - The Islamic months names/abbreviations (Ar&En) are the same in every Arabic country.
 - The weekdays names/abbreviations (Ar&En) are the same in every Arabic country.
 - The Gregorian calendar **English** months names/abbreviations are the same in every Arabic country.
 - The Gregorian calendar **Arabic** months names/abbreviations differ from country to another.

Something I would like to mention here, the translation and the national calendar are two different features. So the end-user should be able to select the calendar he wants (Islamic/Gregorian,..) in any language he prefers.
For example, A user from Egypt may want to have the UI translated to English and the calendar to be displayed as Islamic.

> The examples I've seen show both situations where the Arabic is "primary," meaning the larger text and the Gregorian as the smaller and vice versa.

Yes, it is common to have many calendar types combined into one calendar. In Egypt, we used to have Gregorian, Islamic & Coptic calendars all together as they all are used for different purposes.

**Comment by pkozlowski-opensource** [comment]
We've just landed Hijri calendar support via #1374. This proves that the current implementation of the datepicker has all the flex points necessary to support different calendars.

I'm going to close this issue for now as we've got the basic infrastructure in place. At the same time we don't have skills and plans to implement other calendars - we would expect that other calendar implementations (ex. Hebrew, lunar) will come from the community (very much counting on it!).

Thnx for all that made this happen, we were working towards multiple calendars support for the _long_ time.
